File: src/dev/panels/simulate.ts
/**
 * Driving the state machine and the marker pipeline with no League
 * running.
 *
 * `state_machine::machine`'s pure transition function is well covered by
 * unit tests. `state_machine::supervisor` — the part that spawns watchers,
 * starts the recorder, and writes the finalize row — has almost none, and
 * has never run against a real client. Everything on this page pushes
 * synthetic input through that real code path.
 */
import type { Panel } from "../main";
import { tryCall } from "../ipc";
import {
  card,
  duration,
  escapeHtml,
  kv,
  output,
  panelHead,
  toast,
} from "../ui";
import {
  GAME_STATES,
  type DevSessionView,
  type DispatchReport,
  type FixturesState,
  type InjectReport,
  type ReplayStatus,
} from "../types";

/** The transitions that matter, in the order a real game hits them. */
const EVENTS: Array<{ label: string; event: Record<string, unknown>; note: string }> = [
  {
    label: "Client opened",
    event: { kind: "lockfile_present" },
    note: "Idle → ClientRunning. Spawns a gameflow watcher against a fabricated lockfile, which will fail to connect and retry harmlessly.",
  },
  {
    label: "Phase: ChampSelect",
    event: { kind: "gameflow_phase", phase: "ChampSelect" },
    note: "Not a recording trigger — included so a dodge can be simulated.",
  },
  {
    label: "Phase: InProgress",
    event: { kind: "gameflow_phase", phase: "InProgress" },
    note: "ClientRunning → WaitingForGame.",
  },
  {
    label: "Phase: Reconnect",
    event: { kind: "gameflow_phase", phase: "Reconnect" },
    note: "Treated identically to InProgress — the machine has no memory of how it got here.",
  },
  {
    label: "Live Client up",
    event: { kind: "live_client_up" },
    note: "WaitingForGame → Recording. Really calls Recorder::start.",
  },
  {
    label: "Live Client down",
    event: { kind: "live_client_down" },
    note: "Game crashed mid-match. Recording → Finalizing, preserving whatever was captured.",
  },
  {
    label: "Phase: EndOfGame",
    event: { kind: "gameflow_phase", phase: "EndOfGame" },
    note: "Recording → Finalizing. Really calls Recorder::stop and writes the DB row.",
  },
  {
    label: "Client closed",
    event: { kind: "lockfile_absent" },
    note: "Client crash or quit, from any state.",
  },
];

const DEFAULT_REPLAY_EVENTS = [
  { event_time: 92, event_name: "FirstBlood", Recipient: "Ahri" },
  { event_time: 92, event_name: "ChampionKill", KillerName: "Ahri", VictimName: "Sylas", Assisters: [] },
  { event_time: 260, event_name: "ChampionKill", KillerName: "Viego", VictimName: "Ahri", Assisters: [] },
  { event_time: 420, event_name: "DragonKill", KillerName: "Ahri", DragonType: "Infernal" },
  { event_time: 430, event_name: "ChampionKill", KillerName: "Ahri", VictimName: "Kai'Sa", Assisters: [] },
  { event_time: 436, event_name: "ChampionKill", KillerName: "Ahri", VictimName: "Nami", Assisters: [] },
  { event_time: 441, event_name: "Ace", Acer: "Ahri", AcingTeam: "ORDER" },
  { event_time: 700, event_name: "TurretKilled", KillerName: "Ahri", TurretKilled: "Turret_T2_C_05_A" },
  { event_time: 980, event_name: "BaronKill", KillerName: "Ahri" },
];

let root: HTMLElement | null = null;
let state = "Idle";
let session: DevSessionView | null = null;
let lastDispatch: { value: unknown; error: boolean } | null = null;
let lastInject: { value: unknown; error: boolean } | null = null;
let replay: ReplayStatus | null = null;
let fixtures: FixturesState | null = null;
let snapshotJson = "";
let replayTimer: number | null = null;

function stateStrip(): string {
  return `<div class="states">${GAME_STATES.map(
    (s, i) =>
      `${i > 0 ? `<span class="state-arrow">→</span>` : ""}
       <span class="state-node${s === state ? " current" : ""}">${s}</span>`,
  ).join("")}</div>`;
}

function sessionSummary(): string {
  if (!session) {
    return `<p class="hint">No session open — nothing is collecting markers right now.</p>`;
  }
  return (
    kv([
      ["Markers", String(session.marker_count)],
      ["Samples", String(session.sample_count)],
      ["Elapsed", duration(session.elapsed_s)],
      [
        "Alignment",
        session.alignment_offset_s === null
          ? "not aligned — no snapshot yet"
          : `${session.alignment_offset_s.toFixed(2)}s`,
      ],
    ]) +
    (session.recent_markers.length
      ? `<p class="hint" style="margin-top:.5rem">${escapeHtml(
          session.recent_markers
            .slice(0, 8)
            .map((m) => `${m.kind}@${m.game_time_s.toFixed(0)}s`)
            .join("  ·  "),
        )}</p>`
      : "")
  );
}

function draw() {
  if (!root) return;
  const liveClientFixtures = (fixtures?.entries ?? []).filter((f) =>
    f.group.includes("live-client") || f.name.includes("allgamedata"),
  );

  root.innerHTML =
    panelHead(
      "Simulate",
      "Drive the state machine and the marker pipeline without League running.",
    ) +
    `<div class="warnbar warnbar-danger">
      These are not dry runs. Dispatching an event executes the supervisor's real actions —
      <code>Recorder::start</code> and <code>Recorder::stop</code> included — and a finalize writes
      a real row and a real file.
    </div>` +
    card("Current state", stateStrip() + `<div style="margin-top:.8rem">${sessionSummary()}</div>`) +
    card(
      "State events",
      `<div class="row">${EVENTS.map(
        (e, i) => `<button type="button" data-event="${i}">${escapeHtml(e.label)}</button>`,
      ).join("")}</div>
      <p class="hint-block">A full game is: Client opened → Phase: InProgress → Live Client up →
        (inject snapshots) → Phase: EndOfGame.</p>
      ${lastDispatch ? output(lastDispatch.value, lastDispatch.error) : ""}`,
    ) +
    card(
      "Inject a Live Client Data snapshot",
      `<div class="row" style="margin-bottom:.5rem">
        <select id="fixture-pick">
          <option value="">Load a fixture…</option>
          ${liveClientFixtures
            .map(
              (f) =>
                `<option value="${escapeHtml(f.path)}">${escapeHtml(f.source)} · ${escapeHtml(
                  f.name,
                )}</option>`,
            )
            .join("")}
        </select>
        <button type="button" class="ghost tiny" data-probe>Fetch from a running game</button>
      </div>
      <textarea id="snapshot" rows="10" spellcheck="false"
        placeholder="Paste an allgamedata payload, or load one above">${escapeHtml(snapshotJson)}</textarea>
      <div class="row" style="margin-top:.6rem">
        <button type="button" class="primary" data-inject>Inject</button>
        <span class="hint">Runs the real MarkerTracker and team_diff over this payload.</span>
      </div>
      ${lastInject ? output(lastInject.value, lastInject.error) : ""}`,
    ) +
    card(
      "Scripted replay",
      `<div class="field-grid">
        <label class="field"><span>Game length (s)</span>
          <input type="number" id="replay-duration" value="1200" /></label>
        <label class="field"><span>Speed multiplier</span>
          <input type="number" id="replay-speed" value="60" />
          <span class="hint">60× runs a 20-minute game in 20 seconds</span></label>
      </div>
      <label class="check" style="margin-top:.6rem">
        <input type="checkbox" id="replay-drive" checked />
        Drive the state machine too — start recording, then finalize into a real row and file
      </label>
      <div class="row" style="margin-top:.8rem">
        <button type="button" class="primary" data-replay-start ${
          replay?.running ? "disabled" : ""
        }>Start replay</button>
        <button type="button" class="danger" data-replay-stop ${
          replay?.running ? "" : "disabled"
        }>Stop</button>
        ${
          replay
            ? `<span class="hint num">${replay.game_time_s.toFixed(0)}s / ${replay.duration_s.toFixed(
                0,
              )}s · ${replay.ticks} ticks · ${replay.events_fired} events${
                replay.finished ? " · finished" : ""
              }</span>`
            : ""
        }
      </div>
      ${replay?.error ? output(replay.error, true) : ""}
      <p class="hint-block">Each tick rewrites game time and the event list on the base payload,
        then pushes it through the same <code>on_snapshot</code> the 1 Hz poller uses — so the
        marker tracker's cross-poll de-duplication is exercised too, not bypassed. The base payload
        is whatever is in the snapshot box above; load a fixture first.</p>`,
    ) +
    card(
      "League API probes",
      `<div class="row">
        <label class="field field-inline" style="flex:1 1 22rem">
          <span>LCU path</span>
          <input type="text" id="lcu-path" value="/lol-gameflow/v1/gameflow-phase" style="flex:1" />
        </label>
        <button type="button" data-lcu-get>GET</button>
      </div>
      <div class="row" style="margin-top:.5rem">
        <label class="field field-inline"><span>Game id</span>
          <input type="number" id="game-id" style="width:11rem" /></label>
        <button type="button" data-match-summary>fetch_match_summary</button>
        <span class="hint">Implemented and tested, but called from nowhere in the app — which is
          why every recording's champion, win, and KDA are NULL.</span>
      </div>
      <div id="probe-out"></div>`,
    );
}

/** Null once the panel is unmounted — every caller runs after an await,
 *  so the user may have navigated away in the meantime. */
function textarea(): HTMLTextAreaElement | null {
  return root?.querySelector<HTMLTextAreaElement>("#snapshot") ?? null;
}

async function refreshSession() {
  const [status, s] = await Promise.all([
    tryCall<{ state: string }>("game_state_status"),
    tryCall<DevSessionView | null>("dev_session_snapshot"),
  ]);
  if (status.ok) state = status.value.state;
  session = s.ok ? s.value : null;
}

function pollReplay(on: boolean) {
  if (replayTimer !== null) {
    clearInterval(replayTimer);
    replayTimer = null;
  }
  if (!on) return;
  replayTimer = window.setInterval(async () => {
    const result = await tryCall<ReplayStatus>("dev_replay_status");
    if (!result.ok || !root) return;
    replay = result.value;
    await refreshSession();
    if (!root) return;
    draw();
    if (!replay.running) {
      pollReplay(false);
      if (replay.finished) toast("Replay finished", "ok");
    }
  }, 500);
}

export const simulatePanel: Panel = {
  id: "simulate",
  title: "Simulate",
  icon: "▶",
  group: "Status",

  async mount(el, context) {
    root = el;
    lastDispatch = null;
    lastInject = null;

    // A fixture handed over from the Fixtures panel.
    const handoff = context.payload;
    if (typeof handoff === "string") snapshotJson = handoff;

    const [f, r] = await Promise.all([
      tryCall<FixturesState>("dev_fixtures_state"),
      tryCall<ReplayStatus>("dev_replay_status"),
    ]);
    fixtures = f.ok ? f.value : null;
    replay = r.ok ? r.value : null;
    await refreshSession();
    draw();
    if (replay?.running) pollReplay(true);

    el.addEventListener("change", async (e) => {
      const picker = (e.target as HTMLElement).closest<HTMLSelectElement>("#fixture-pick");
      if (!picker || !picker.value) return;
      const result = await tryCall<string>("dev_fixture_read", { path: picker.value });
      const box = textarea();
      if (!box) return;
      if (result.ok) {
        snapshotJson = result.value;
        box.value = result.value;
        toast("Fixture loaded", "ok");
      } else {
        toast(result.error, "err");
      }
    });

    el.addEventListener("click", async (e) => {
      const target = e.target as HTMLElement;

      const eventBtn = target.closest<HTMLElement>("[data-event]");
      if (eventBtn) {
        const spec = EVENTS[Number(eventBtn.dataset.event)];
        const result = await tryCall<DispatchReport>("dev_dispatch_state_event", {
          event: spec.event,
        });
        if (!root) return;
        if (result.ok) {
          const r = result.value;
          state = r.after.state;
          session = r.session;
          lastDispatch = {
            value:
              r.before.state === r.after.state
                ? `${r.before.state} → (no change). ${spec.note}`
                : `${r.before.state} → ${r.after.state}. ${spec.note}`,
            error: false,
          };
        } else {
          lastDispatch = { value: result.error, error: true };
        }
        draw();
        return;
      }

      if (target.closest("[data-probe]")) {
        const result = await tryCall<unknown>("dev_live_client_probe");
        const box = textarea();
        if (!box) return;
        if (result.ok) {
          snapshotJson = JSON.stringify(result.value, null, 2);
          box.value = snapshotJson;
          toast("Fetched live snapshot", "ok");
        } else {
          toast(result.error, "err");
        }
        return;
      }

      if (target.closest("[data-inject]")) {
        const box = textarea();
        if (!box) return;
        snapshotJson = box.value;
        let parsed: unknown;
        try {
          parsed = JSON.parse(snapshotJson);
        } catch (err) {
          lastInject = { value: `Not valid JSON: ${err}`, error: true };
          draw();
          return;
        }
        const result = await tryCall<InjectReport>("dev_inject_snapshot", { snapshot: parsed });
        if (!root) return;
        if (result.ok) {
          const r = result.value;
          session = r.session;
          lastInject = {
            value: r.accepted
              ? `Accepted in state ${r.state}: +${r.markers_added} marker(s), +${r.samples_added} sample(s).`
              : r.note,
            error: !r.accepted,
          };
        } else {
          lastInject = { value: result.error, error: true };
        }
        await refreshSession();
        draw();
        return;
      }

      if (target.closest("[data-replay-start]")) {
        const box = textarea();
        if (!box) return;
        snapshotJson = box.value;
        let base: unknown;
        try {
          base = JSON.parse(snapshotJson);
        } catch {
          toast("Load a base snapshot into the box above first", "err");
          return;
        }
        const num = (id: string, fallback: number) =>
          Number(root?.querySelector<HTMLInputElement>(id)?.value ?? fallback);
        const spec = {
          base_snapshot: base,
          duration_s: num("#replay-duration", 1200),
          speed: num("#replay-speed", 60),
          events: DEFAULT_REPLAY_EVENTS,
          drive_state_machine:
            root?.querySelector<HTMLInputElement>("#replay-drive")?.checked ?? true,
        };
        const result = await tryCall("dev_replay_start", { spec });
        if (result.ok) {
          toast("Replay started");
          pollReplay(true);
        } else {
          toast(result.error, "err");
        }
        return;
      }

      if (target.closest("[data-replay-stop]")) {
        await tryCall("dev_replay_stop");
        pollReplay(false);
        replay = null;
        await refreshSession();
        draw();
        return;
      }

      if (target.closest("[data-lcu-get]")) {
        const path = root?.querySelector<HTMLInputElement>("#lcu-path")?.value.trim() ?? "";
        const result = await tryCall<unknown>("dev_lcu_get", { path });
        const out = root?.querySelector<HTMLElement>("#probe-out");
        if (out) out.innerHTML = output(result.ok ? result.value : result.error, !result.ok);
        return;
      }

      if (target.closest("[data-match-summary]")) {
        const gameId = Number(root?.querySelector<HTMLInputElement>("#game-id")?.value);
        if (Number.isNaN(gameId)) {
          toast("Enter a game id", "err");
          return;
        }
        const result = await tryCall<unknown>("dev_fetch_match_summary", { gameId });
        const out = root?.querySelector<HTMLElement>("#probe-out");
        if (out) out.innerHTML = output(result.ok ? result.value : result.error, !result.ok);
      }
    });
  },

  onHealth(health) {
    // Keep the strip live without redrawing the page under the user's
    // cursor — a full redraw would blow away the snapshot textarea.
    if (health.supervisor.state !== state) {
      state = health.supervisor.state;
      session = health.session;
      const strip = root?.querySelector<HTMLElement>(".states");
      if (strip) strip.outerHTML = stateStrip();
    }
  },

  unmount() {
    pollReplay(false);
    root = null;
  },
};

File: src-tauri/src/dev/simulate.rs
//! Driving the state machine and the marker pipeline without League.
//!
//! `state_machine::machine`'s pure transition function has eleven unit
//! tests; `state_machine::supervisor` — the part that actually spawns
//! watchers, starts the recorder, and writes the finalize row — has two,
//! and has never run against a real LCU (DEVELOPMENT.md §3.4). Everything
//! here feeds that untested glue synthetic input through its real code
//! path: no mocks, no dry runs. Dispatching `GameflowPhase(InProgress)`
//! really does start the recorder.

use crate::live_client::AllGameData;
use crate::state_machine::{DevSessionView, StateEvent, SupervisorStatus};
use crate::{dev, lcu, AppState};
use serde::{Deserialize, Serialize};
use std::sync::{Arc, Mutex};
use std::time::Duration;
use tauri::async_runtime::JoinHandle;

/// JSON-friendly mirror of `StateEvent`. The real enum carries
/// `LockfileState`/`GameflowPhase`, neither of which deserializes from
/// anything a form can produce — `GameflowPhase` in particular has a
/// hand-written `Deserialize` that takes a bare string, and `Unknown` is
/// its catch-all, so a typo would silently become a valid-but-inert phase
/// rather than an error the portal can show.
#[derive(Debug, Clone, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum DevStateEvent {
    /// League Client appeared. Port and password are fabricated by default
    /// — the state machine only stores them to hand to the gameflow
    /// watcher, which will fail to connect and retry harmlessly.
    LockfilePresent {
        #[serde(default)]
        port: Option<u16>,
        #[serde(default)]
        password: Option<String>,
    },
    LockfileAbsent,
    GameflowPhase {
        phase: String,
    },
    LiveClientUp,
    LiveClientDown,
    FinalizeComplete,
}

impl DevStateEvent {
    fn into_state_event(self) -> StateEvent {
        match self {
            DevStateEvent::LockfilePresent { port, password } => {
                StateEvent::LockfileChanged(lcu::LockfileState::Present(lcu::LockfileInfo {
                    name: "LeagueClient".to_string(),
                    pid: 0,
                    port: port.unwrap_or(2999),
                    password: password.unwrap_or_else(|| "dev-portal".to_string()),
                    protocol: "https".to_string(),
                }))
            }
            DevStateEvent::LockfileAbsent => StateEvent::LockfileChanged(lcu::LockfileState::Absent),
            DevStateEvent::GameflowPhase { phase } => {
                StateEvent::GameflowPhase(lcu::GameflowPhase::from(phase.as_str()))
            }
            DevStateEvent::LiveClientUp => StateEvent::LiveClientUp,
            DevStateEvent::LiveClientDown => StateEvent::LiveClientDown,
            DevStateEvent::FinalizeComplete => StateEvent::FinalizeComplete,
        }
    }
}

#[derive(Serialize)]
pub struct DispatchReport {
    pub before: SupervisorStatus,
    pub after: SupervisorStatus,
    pub session: Option<DevSessionView>,
}

/// Feeds one event through the live supervisor. Reports the state on
/// either side so the portal can show the transition rather than just the
/// result — a dispatch that changes nothing is the interesting case.
#[tauri::command]
pub fn dev_dispatch_state_event(
    state: tauri::State<AppState>,
    app: tauri::AppHandle,
    event: DevStateEvent,
) -> Result<DispatchReport, String> {
    let before = state.supervisor.status();
    state.supervisor.dev_dispatch(event.into_state_event());
    let after = state.supervisor.status();

    // A transition out of Recording writes a row; anything else can't.
    if before.state != after.state {
        dev::notify_library_changed(&app);
    }

    Ok(DispatchReport {
        session: state.supervisor.dev_session_view(),
        before,
        after,
    })
}

#[derive(Serialize)]
pub struct InjectReport {
    pub accepted: bool,
    /// Why nothing happened, when `accepted` is false. Injecting outside a
    /// recording is the overwhelmingly common mistake, and it looks
    /// identical to a broken pipeline unless it's said out loud.
    pub note: Option<String>,
    pub markers_added: usize,
    pub samples_added: usize,
    pub session: Option<DevSessionView>,
    pub state: String,
}

/// Feeds one Live Client Data payload through the real marker and sample
/// pipeline, exactly as the 1 Hz poller would.
#[tauri::command]
pub fn dev_inject_snapshot(
    state: tauri::State<AppState>,
    snapshot: serde_json::Value,
) -> Result<InjectReport, String> {
    let parsed: AllGameData = serde_json::from_value(snapshot)
        .map_err(|e| format!("not a valid Live Client Data payload: {e}"))?;

    let before = state.supervisor.dev_session_view();
    state.supervisor.dev_on_snapshot(parsed);
    let after = state.supervisor.dev_session_view();

    let (markers_added, samples_added) = match (&before, &after) {
        (Some(b), Some(a)) => (
            a.marker_count.saturating_sub(b.marker_count),
            a.sample_count.saturating_sub(b.sample_count),
        ),
        (None, Some(a)) => (a.marker_count, a.sample_count),
        _ => (0, 0),
    };

    let status = state.supervisor.status();
    Ok(InjectReport {
        accepted: after.is_some(),
        note: after.is_none().then(|| {
            "No recording session is open, so the snapshot was discarded. Dispatch \
             LockfilePresent → GameflowPhase(InProgress) first — the supervisor only \
             collects markers between Recorder::start and Recorder::stop."
                .to_string()
        }),
        markers_added,
        samples_added,
        session: after,
        state: format!("{:?}", status.state),
    })
}

/// The in-flight session — markers and samples accumulating right now.
/// `game_state_status` only carries the *last finalized* recording, so
/// without this there is no way to watch the pipeline work.
#[tauri::command]
pub fn dev_session_snapshot(state: tauri::State<AppState>) -> Option<DevSessionView> {
    state.supervisor.dev_session_view()
}

// --- Scripted replay ---------------------------------------------------

#[derive(Debug, Clone, Deserialize)]
pub struct ReplaySpec {
    /// Base Live Client Data payload. Its `gameData.gameTime` and
    /// `events.Events` are rewritten each tick; everything else
    /// (`allPlayers`, `activePlayer`) is passed through, which is what
    /// makes `team_diff` produce a real curve.
    pub base_snapshot: serde_json::Value,
    /// Game seconds to simulate.
    pub duration_s: f64,
    /// Wall-clock speed multiplier. 60 means a 20-minute game finishes in
    /// 20 seconds.
    pub speed: f64,
    /// Scripted events spliced into the stream as game time passes them.
    #[serde(default)]
    pub events: Vec<ReplayEvent>,
    /// Dispatch the state events that start and stop a recording around
    /// the replay, so it finalizes into a real DB row and file.
    #[serde(default)]
    pub drive_state_machine: bool,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ReplayEvent {
    pub event_time: f64,
    pub event_name: String,
    #[serde(default, flatten)]
    pub extra: serde_json::Map<String, serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Default)]
pub struct ReplayStatus {
    pub running: bool,
    pub game_time_s: f64,
    pub duration_s: f64,
    pub ticks: u64,
    pub events_fired: usize,
    pub finished: bool,
    pub error: Option<String>,
}

pub struct ReplayHandle {
    pub(crate) task: JoinHandle<()>,
    pub(crate) status: Arc<Mutex<ReplayStatus>>,
}

impl Drop for ReplayHandle {
    fn drop(&mut self) {
        self.task.abort();
    }
}

/// Plays a whole fake game. Each tick rewrites `gameTime` and the event
/// list, then pushes the payload through the same `on_snapshot` the poller
/// uses — so the `MarkerTracker`'s cross-poll de-duplication (each real
/// poll returns the *entire* event history, not just new events) is
/// exercised too, not bypassed.
#[tauri::command]
pub fn dev_replay_start(
    state: tauri::State<AppState>,
    dev: tauri::State<super::DevState>,
    app: tauri::AppHandle,
    spec: ReplaySpec,
) -> Result<(), String> {
    let mut slot = dev.replay.lock().map_err(|e| e.to_string())?;
    if slot.is_some() {
        return Err("a replay is already running — stop it first".to_string());
    }

    // Validate up front: a malformed base payload should fail the button
    // press, not a background task nobody is watching.
    serde_json::from_value::<AllGameData>(spec.base_snapshot.clone())
        .map_err(|e| format!("base_snapshot is not a valid Live Client Data payload: {e}"))?;
    if !spec.base_snapshot.is_object() {
        return Err("base_snapshot must be a JSON object".to_string());
    }
    let speed = if spec.speed > 0.0 { spec.speed } else { 1.0 };

    let status = Arc::new(Mutex::new(ReplayStatus {
        running: true,
        duration_s: spec.duration_s,
        ..Default::default()
    }));

    let supervisor = Arc::clone(&state.supervisor);
    let task_status = Arc::clone(&status);
    let app_handle = app.clone();

    let task = tauri::async_runtime::spawn(async move {
        if spec.drive_state_machine {
            supervisor.dev_dispatch(StateEvent::LockfileChanged(lcu::LockfileState::Present(
                lcu::LockfileInfo {
                    name: "LeagueClient".to_string(),
                    pid: 0,
                    port: 2999,
                    password: "dev-portal".to_string(),
                    protocol: "https".to_string(),
                },
            )));
            supervisor.dev_dispatch(StateEvent::GameflowPhase(lcu::GameflowPhase::InProgress));
        }

        // One tick per game second, wall-clock-scaled. Matching the real
        // poller's 1 Hz keeps the sample series the same density a live
        // game produces.
        let tick = Duration::from_secs_f64((1.0 / speed).clamp(0.001, 5.0));
        let total = spec.duration_s.max(0.0) as u64;

        for second in 0..=total {
            let game_time_s = second as f64;
            let fired: Vec<&ReplayEvent> = spec
                .events
                .iter()
                .filter(|e| e.event_time <= game_time_s)
                .collect();

            let mut payload = spec.base_snapshot.clone();
            if let Some(obj) = payload.as_object_mut() {
                obj.insert(
                    "gameData".to_string(),
                    serde_json::json!({ "gameTime": game_time_s }),
                );
                obj.insert(
                    "events".to_string(),
                    serde_json::json!({
                        "Events": fired
                            .iter()
                            .enumerate()
                            .map(|(i, e)| {
                                let mut v = serde_json::Map::new();
                                v.insert("EventID".to_string(), serde_json::json!(i));
                                v.insert("EventName".to_string(), serde_json::json!(e.event_name));
                                v.insert("EventTime".to_string(), serde_json::json!(e.event_time));
                                for (k, val) in &e.extra {
                                    v.insert(k.clone(), val.clone());
                                }
                                serde_json::Value::Object(v)
                            })
                            .collect::<Vec<_>>()
                    }),
                );
            }

            match serde_json::from_value::<AllGameData>(payload) {
                Ok(snapshot) => supervisor.dev_on_snapshot(snapshot),
                Err(e) => {
                    let mut s = task_status.lock().unwrap();
                    s.error = Some(e.to_string());
                    s.running = false;
                    return;
                }
            }

            {
                let mut s = task_status.lock().unwrap();
                s.game_time_s = game_time_s;
                s.ticks += 1;
                s.events_fired = fired.len();
            }

            tokio::time::sleep(tick).await;
        }

        if spec.drive_state_machine {
            supervisor.dev_dispatch(StateEvent::GameflowPhase(lcu::GameflowPhase::EndOfGame));
            supervisor.dev_emit_library_changed();
        }
        dev::notify_library_changed(&app_handle);

        let mut s = task_status.lock().unwrap();
        s.running = false;
        s.finished = true;
    });

    *slot = Some(ReplayHandle { task, status });
    Ok(())
}

#[tauri::command]
pub fn dev_replay_stop(dev: tauri::State<super::DevState>) -> Result<(), String> {
    // `ReplayHandle::drop` aborts the task.
    dev.replay.lock().map_err(|e| e.to_string())?.take();
    Ok(())
}

#[tauri::command]
pub fn dev_replay_status(dev: tauri::State<super::DevState>) -> Result<ReplayStatus, String> {
    let slot = dev.replay.lock().map_err(|e| e.to_string())?;
    Ok(match slot.as_ref() {
        Some(handle) => handle.status.lock().map_err(|e| e.to_string())?.clone(),
        None => ReplayStatus::default(),
    })
}

// --- Live API probes ---------------------------------------------------

/// Raw GET against any LCU path, so an endpoint can be inspected before
/// any parsing code is written for it. Returns the response as JSON.
#[tauri::command]
pub async fn dev_lcu_get(path: String) -> Result<serde_json::Value, String> {
    let lockfile = lcu::lockfile::discover()
        .map_err(|e| e.to_string())?
        .ok_or_else(|| "League Client not running (no lockfile found)".to_string())?;
    let client = lcu::LcuHttpClient::new(&lockfile).map_err(|e| e.to_string())?;
    client
        .get_json::<serde_json::Value>(&path)
        .await
        .map_err(|e| e.to_string())
}

/// Exercises `lcu::match_data::fetch_match_summary`, which is fully
/// implemented and unit-tested but called from nowhere in the app — which
/// is why every `RecordingRow`'s `champion`/`win`/`kda_*` is NULL in
/// practice. Wiring it into finalize is a separate change; this at least
/// makes it runnable against a real client.
#[tauri::command]
pub async fn dev_fetch_match_summary(game_id: i64) -> Result<lcu::MatchSummary, String> {
    let lockfile = lcu::lockfile::discover()
        .map_err(|e| e.to_string())?
        .ok_or_else(|| "League Client not running (no lockfile found)".to_string())?;
    let client = lcu::LcuHttpClient::new(&lockfile).map_err(|e| e.to_string())?;
    lcu::fetch_match_summary(&client, game_id)
        .await
        .map_err(|e| e.to_string())
}

/// One-shot fetch from the in-game Live Client Data API, returned raw so
/// it can be saved as a fixture.
#[tauri::command]
pub async fn dev_live_client_probe() -> Result<serde_json::Value, String> {
    let client = crate::live_client::LiveClientDataClient::new().map_err(|e| e.to_string())?;
    client
        .fetch_all_game_data_raw()
        .await
        .map_err(|e| e.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn parse(json: &str) -> DevStateEvent {
        serde_json::from_str(json).expect("valid dev state event")
    }

    #[test]
    fn maps_lockfile_events() {
        let StateEvent::LockfileChanged(lcu::LockfileState::Present(info)) =
            parse(r#"{"kind":"lockfile_present","port":1234}"#).into_state_event()
        else {
            panic!("expected a present lockfile");
        };
        assert_eq!(info.port, 1234);

        assert!(matches!(
            parse(r#"{"kind":"lockfile_absent"}"#).into_state_event(),
            StateEvent::LockfileChanged(lcu::LockfileState::Absent)
        ));
    }

    #[test]
    fn maps_known_gameflow_phases() {
        let StateEvent::GameflowPhase(phase) =
            parse(r#"{"kind":"gameflow_phase","phase":"InProgress"}"#).into_state_event()
        else {
            panic!("expected a gameflow phase");
        };
        assert_eq!(phase, lcu::GameflowPhase::InProgress);
    }

    /// A typo must stay visible as `Unknown(..)` rather than silently
    /// becoming a phase the state machine ignores — the portal shows the
    /// resulting state, and "nothing happened" needs an explanation.
    #[test]
    fn unknown_phase_names_round_trip_verbatim() {
        let StateEvent::GameflowPhase(phase) =
            parse(r#"{"kind":"gameflow_phase","phase":"InProgres"}"#).into_state_event()
        else {
            panic!("expected a gameflow phase");
        };
        assert_eq!(phase, lcu::GameflowPhase::Unknown("InProgres".to_string()));
    }

    #[test]
    fn maps_live_client_and_finalize_events() {
        assert!(matches!(
            parse(r#"{"kind":"live_client_up"}"#).into_state_event(),
            StateEvent::LiveClientUp
        ));
        assert!(matches!(
            parse(r#"{"kind":"live_client_down"}"#).into_state_event(),
            StateEvent::LiveClientDown
        ));
        assert!(matches!(
            parse(r#"{"kind":"finalize_complete"}"#).into_state_event(),
            StateEvent::FinalizeComplete
        ));
    }
}
